File: VignetteBackend/infrastructure/docker/Dockerfile.scala
# Multi-stage Dockerfile for Scala services
# Stage 1: Build
FROM hseeberger/scala-sbt:11.0.16.1_1.8.2_2.13.10 AS builder

WORKDIR /app

# Copy build files
COPY build.sbt .
COPY project ./project

# Copy source code
COPY src ./src

# Build the application
RUN sbt clean assembly

# Stage 2: Runtime
FROM openjdk:11-jre-slim

WORKDIR /root/

# Copy JAR from builder
COPY --from=builder /app/target/scala-*/story-service-assembly-*.jar ./app.jar

# Expose ports
EXPOSE 8090 50005

# Health check
HEALTHCHECK --interval=30s --timeout=3s --start-period=10s --retries=3 \
  CMD wget --no-verbose --tries=1 --spider http://localhost:${PORT:-8090}/health || exit 1

# Run the application
CMD ["java", "-jar", "./app.jar"]

File: SocialinkBackend/infrastructure/docker/Dockerfile.go
# Multi-stage Dockerfile for Go services
# Stage 1: Build
FROM golang:1.21-alpine AS builder

# Install build dependencies
RUN apk add --no-cache git gcc musl-dev

WORKDIR /app

# Copy go mod files
COPY go.mod go.sum ./

# Download dependencies
RUN go mod download

# Copy source code
COPY . .

# Build the application
RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o main ./cmd/api

# Stage 2: Runtime
FROM alpine:latest

# Install ca-certificates for HTTPS
RUN apk --no-cache add ca-certificates

WORKDIR /root/

# Copy binary from builder
COPY --from=builder /app/main .

# Copy migrations (if exists)
COPY --from=builder /app/migrations ./migrations 2>/dev/null || true

# Expose ports (HTTP and gRPC)
EXPOSE 8080 50051

# Health check
HEALTHCHECK --interval=30s --timeout=3s --start-period=5s --retries=3 \
  CMD wget --no-verbose --tries=1 --spider http://localhost:${PORT:-8080}/health || exit 1

# Run the application
CMD ["./main"]
